TiFlash segmentation fault

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
partition table r/w and ddl
### 2. What did you expect to see? (Required)
```
有一个 partition ddl 的 case ，tiflash crash 了 @陈桐立
 /cc
[2022/07/15 06:09:40.896 +08:00] [ERROR] [BaseDaemon.cpp:377] [BaseDaemon:########################################] [thread_id=2637]
[2022/07/15 06:09:40.896 +08:00] [ERROR] [BaseDaemon.cpp:378] ["BaseDaemon:(from thread 2056) Received signal Segmentation fault(11)."] [thread_id=2637]
[2022/07/15 06:09:40.896 +08:00] [ERROR] [BaseDaemon.cpp:408] ["BaseDaemon:Address: 0x28"] [thread_id=2637]
[2022/07/15 06:09:40.896 +08:00] [ERROR] [BaseDaemon.cpp:414] ["BaseDaemon:Access: read."] [thread_id=2637]
[2022/07/15 06:09:40.896 +08:00] [ERROR] [BaseDaemon.cpp:423] ["BaseDaemon:Address not mapped to object."] [thread_id=2637]
[2022/07/15 06:09:40.900 +08:00] [ERROR] [BaseDaemon.cpp:570] ["BaseDaemon:
       0x6594971\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+106514801]
                \tlibs/libdaemon/src/BaseDaemon.cpp:221
  0x7fe54c045630\t<unknown symbol> [libpthread.so.0+63024]
       0x75e22ad\tDB::TunnelSender::getConsumerFinishMsg() [tiflash+123609773]
                \tdbms/src/Flash/Mpp/MPPTunnel.h:81
       0x75e2171\tDB::MPPTunnel::write(mpp::MPPDataPacket const&, bool) [tiflash+123609457]
                \tdbms/src/Flash/Mpp/MPPTunnel.cpp:153
       0x1cae2a4\tDB::MPPTunnelSetBase<DB::MPPTunnel>::write(tipb::SelectResponse&) [tiflash+30073508]
                \tdbms/src/Flash/Mpp/MPPTunnelSet.cpp:64
       0x75a15d6\tDB::StreamingDAGResponseWriter<std::__1::shared_ptr<DB::MPPTunnelSet>, false>::finishWrite() [tiflash+123344342]
                \tdbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.cpp:92
       0x75845e4\tDB::ExchangeSenderBlockInputStream::readSuffixImpl() [tiflash+123225572]
                \tdbms/src/DataStreams/ExchangeSenderBlockInputStream.h:45
       0x65be988\tDB::IProfilingBlockInputStream::readSuffix() [tiflash+106686856]
                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:141
       0x7588b5a\tDB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::readSuffix() [tiflash+123243354]
                \tdbms/src/DataStreams/UnionBlockInputStream.h:255
       0x75d223b\tDB::MPPTask::runImpl() [tiflash+123544123]
                \tdbms/src/Flash/Mpp/MPPTask.cpp:371
       0x1d7c895\tauto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()::operator()() [tiflash+30918805]
                \tdbms/src/Common/wrapInvocable.h:36
       0x1d80243\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+30933571]
                \tdbms/src/Common/DynamicThreadPool.cpp:101
       0x1d7ff16\tDB::DynamicThreadPool::dynamicWork(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >) [tiflash+30932758]
                \tdbms/src/Common/DynamicThreadPool.cpp:142
       0x1d8192b\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > > >(void*) [tiflash+30939435]
                \t/usr/local/bin/../include/c++/v1/thread:291
  0x7fe54c03dea5\tstart_thread [libpthread.so.0+32421]"] [thread_id=2637]
```
### 3. What did you see instead (Required)

### 4. What is your TiFlash version? (Required)
master

--- a/dbms/src/Flash/Mpp/MPPTunnel.cpp
+++ b/dbms/src/Flash/Mpp/MPPTunnel.cpp
@@ -150,7 +150,7 @@ void MPPTunnel::write(const mpp::MPPDataPacket & data, bool close_after_write)
             std::unique_lock lk(mu);
             waitUntilConnectedOrFinished(lk);
             if (status == TunnelStatus::Finished)
-                throw Exception("write to tunnel which is already closed," + tunnel_sender->getConsumerFinishMsg());
+                throw Exception(fmt::format("write to tunnel which is already closed,{}", tunnel_sender ? tunnel_sender->getConsumerFinishMsg() : ""));
         }
 
         if (send_queue->push(std::make_shared<mpp::MPPDataPacket>(data)))
@@ -178,7 +178,7 @@ void MPPTunnel::writeDone()
     {
         std::unique_lock lk(mu);
         if (status == TunnelStatus::Finished)
-            throw Exception("write to tunnel which is already closed," + tunnel_sender->getConsumerFinishMsg());
+            throw Exception(fmt::format("write to tunnel which is already closed,{}", tunnel_sender ? tunnel_sender->getConsumerFinishMsg() : ""));
         /// make sure to finish the tunnel after it is connected
         waitUntilConnectedOrFinished(lk);
         finishSendQueue();
@@ -238,6 +238,10 @@ void MPPTunnel::waitForSenderFinish(bool allow_throw)
     LOG_FMT_TRACE(log, "start wait for consumer finish!");
     {
         std::unique_lock lock(mu);
+        if (status == TunnelStatus::Finished)
+        {
+            return;
+        }
         status = TunnelStatus::WaitingForSenderFinish;
     }
     String err_msg = tunnel_sender->getConsumerFinishMsg(); // may blocking

--- a/dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp
+++ b/dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp
@@ -301,6 +301,38 @@ try
 }
 CATCH
 
+TEST_F(TestMPPTunnel, WriteAfterUnconnectFinished)
+{
+    try
+    {
+        auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+        std::unique_ptr<mpp::MPPDataPacket> data_packet_ptr = std::make_unique<mpp::MPPDataPacket>();
+        setTunnelFinished(mpp_tunnel_ptr);
+        data_packet_ptr->set_data("First");
+        mpp_tunnel_ptr->write(*data_packet_ptr);
+        GTEST_FAIL();
+    }
+    catch (Exception & e)
+    {
+        GTEST_ASSERT_EQ(e.message(), "write to tunnel which is already closed,");
+    }
+}
+
+TEST_F(TestMPPTunnel, WriteDoneAfterUnconnectFinished)
+{
+    try
+    {
+        auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+        setTunnelFinished(mpp_tunnel_ptr);
+        mpp_tunnel_ptr->writeDone();
+        GTEST_FAIL();
+    }
+    catch (Exception & e)
+    {
+        GTEST_ASSERT_EQ(e.message(), "write to tunnel which is already closed,");
+    }
+}
+
 TEST_F(TestMPPTunnel, ConnectWriteCancel)
 try
 {

